flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java

```diff
@@ -0,0 +1,251 @@
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.configuration.BlobServerOptions;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.util.concurrent.FutureUtils;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import javax.annotation.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.apache.flink.runtime.blob.BlobServerPutTest.put;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Test for using {@link BlobCacheSizeTracker} to track the size of BLOBs in {@link
+ * PermanentBlobCache}. When new BLOBs are going to be inserted and the size limit exceeds, {@link
+ * BlobCacheSizeTracker} will provide excess BLOBs for {@link PermanentBlobCache} to delete.
+ */
+public class PermanentBlobCacheSizeLimitTest {
+
+    private static final int BLOB_SIZE = 10_000;
+
+    private static final Random RANDOM = new Random();
+
+    @Rule public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @Test
+    public void testTrackSizeLimitAndDeleteExcess() throws Exception {
+
+        final BlobKey.BlobType blobType = BlobKey.BlobType.PERMANENT_BLOB;
+
+        final Configuration config = new Configuration();
+        config.setString(
+                BlobServerOptions.STORAGE_DIRECTORY, temporaryFolder.newFolder().getAbsolutePath());
+
+        // The size limit is the size of 2 blobs
+        int sizeLimit = 2;
+
+        try (BlobServer server = new BlobServer(config, new VoidBlobStore());
+                BlobCacheService cache =
+                        initBlobCacheServiceWithSizeLimit(
+                                config,
+                                new VoidBlobStore(),
+                                new InetSocketAddress("localhost", server.getPort()),
+                                sizeLimit * BLOB_SIZE)) {
+
+            server.start();
+
+            // Initialize the data structures
+            JobID[] jobId = new JobID[4];
+            byte[][] data = new byte[4][BLOB_SIZE];
+            BlobKey[] key = new BlobKey[4];
+            File[] file = new File[4];
+
+            // Prepare the blob data
+            byte[] blobData = new byte[BLOB_SIZE];
+            RANDOM.nextBytes(blobData);
+
+            for (int i = 0; i < 4; i++) {
+                jobId[i] = new JobID();
+                data[i] = Arrays.copyOf(blobData, blobData.length);
+                data[i][i] ^= 1;
```

Comment:
Maybe add a comments to say that this is to make contents different from each other for later verification?